compiler/rustc_resolve/src/late/diagnostics.rs

```diff
@@ -1810,18 +1810,27 @@ impl<'a: 'ast, 'ast> LateResolutionVisitor<'a, '_, 'ast> {
         false
     }
 
-    fn let_binding_suggestion(&self, err: &mut Diagnostic, ident_span: Span) -> bool {
+    fn let_binding_suggestion(&mut self, err: &mut Diagnostic, ident_span: Span) -> bool {
         // try to give a suggestion for this pattern: `name = 1`, which is common in other languages
         let mut added_suggestion = false;
-        if let Some(Expr { kind: ExprKind::Assign(lhs, _rhs, _), .. }) = self.diagnostic_metadata.in_assignment &&
-            let ast::ExprKind::Path(None, _) = lhs.kind {
+        if let Some(Expr { kind: ExprKind::Assign(lhs, rhs, _), .. }) =
+            self.diagnostic_metadata.in_assignment
+        {
+            let is_rhs_assign = match rhs.kind {
+                ExprKind::Assign(..) => true,
+                _ => false,
+            };
+
+            if let ast::ExprKind::Path(None, _) = lhs.kind && !is_rhs_assign {
                 let sm = self.r.session.source_map();
                 let line_span = sm.span_extend_to_line(ident_span);
                 let ident_name = sm.span_to_snippet(ident_span).unwrap();
-                // HACK(chenyukang): make sure ident_name is at the starting of the line to protect against macros
-                if sm
-                    .span_to_snippet(line_span)
-                    .map_or(false, |s| s.trim().starts_with(&ident_name))
+                // HACK(chenyukang): make sure ident_name is at the starting of the line to protect against macros,
+                // and avoid some special cases like `x = x = x`
+                if let Ok(line) = sm.span_to_snippet(line_span) &&
+                    let stripped = line.split_whitespace().collect::<String>() &&
+                    stripped.trim().starts_with(&ident_name) &&
+                    stripped.matches(&format!("{}=", &ident_name)).count() == 1
```

Comment:
is this really necessary? We could just propose the `let` everywhere.

Reply:
This is weird 😂
![image](https://user-images.githubusercontent.com/230646/200815112-5ecceca1-fa1a-4131-969d-fef410b47764.png)

Reply:
I think we should avoid the entire `source_map` usage. It is very fragile. Maybe try using `tcx.hir().get_parent_node(hir_id)` and check if that is an assignment and bail out then.

Reply:
It's name resolving phase here, we haven't built `hir` related data structures...

Reply:
As far as I know, we have `parent_scope`.
To make life easier for diagnostics, we have this to store some info when walking the AST.

https://github.com/rust-lang/rust/blob/0aaad9e757207657064d70bf9e6c1e6eb327bf15/compiler/rustc_resolve/src/late.rs#L490

It's common to use `source_map` in diagnostics in parsing and resolving phase, but we should limited it for the usage of reducing some false tip or useless diagnostics.

Reply:
I tried to introduced a new variable `is_assign_rhs`, it works as expected.

But we still can not avoid `source_map` totally, for example this code:

```rust
 while x = blah {
 ....
}
```

We don't want to suggest to `while let x = blah`.

And diagnostics reporting is on error handling path, but add `is_assign_rhs` is on normal code path, so I'm not sure whether it's worth.

Maybe we should tolerate ugly code for better performance? 😁
Anyway, we are handling this kind of special werid case:
`x = x = x = y`